stock_move_manual_lot/tests/test_stock_move_manual_lot.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2021 Hunki Enterprises BV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo.exceptions import UserError
+from odoo.tests.common import TransactionCase
+
+
+class TestStockMoveManualLot(TransactionCase):
+    def setUp(self):
+        super().setUp()
+        self.product = self.env['product.product'].create({
+            'name': 'Tracked product',
+            'tracking': 'serial',
+            'type': 'product',
+        })
+        self.lot1 = self.env['stock.production.lot'].create({
+            'name': 'lot1',
+            'product_id': self.product.id,
+        })
+        self.lot2 = self.env['stock.production.lot'].create({
+            'name': 'lot2',
+            'product_id': self.product.id,
+        })
+        self.picking = self.env['stock.picking'].create({
+            'picking_type_id': self.env.ref('stock.picking_type_out').id,
+            'location_id': self.env.ref('stock.stock_location_stock').id,
+            'location_dest_id':
+            self.env.ref('stock.stock_location_customers').id,
+            'move_lines': [(0, 0, {
+                'name': 'testmove',
+                'product_id': self.product.id,
+                'product_uom': self.product.uom_id.id,
+                'product_uom_qty': 1,
+            })],
+        })
+        self.quant = self.env['stock.quant'].create({
+            'product_id': self.product.id,
+            'quantity': 1,
+            'location_id': self.picking.location_id.id,
+            'lot_id': self.lot1.id,
+        })
+        self.picking.action_assign()
+        self.quant = self.env['stock.quant'].create({
+            'product_id': self.product.id,
+            'quantity': 1,
+            'location_id': self.picking.location_id.id,
+            'lot_id': self.lot2.id,
+        })
+        self.picking.move_line_ids.qty_done =\
+            self.picking.move_line_ids.product_qty
+
+    def test_force_manual_selection(self):
+        with self.assertRaises(UserError), self.env.clear_upon_failure():
+            self.picking.button_validate()
+        self.picking.move_line_ids.manual_lot_id = self.lot1
+        self.picking.button_validate()
+        self.assertEqual(self.picking.state, 'done')
+        self.assertEqual(self.env['stock.quant'].search([
+            ('location_id', '=', self.picking.location_dest_id.id),
+            ('product_id', '=', self.product.id),
+        ]).mapped('lot_id'), self.lot1)
+
+    def test_reassign_reservation(self):
+        picking2 = self.picking.copy()
+        picking2.action_assign()
+        self.assertEqual(picking2.move_line_ids.lot_id, self.lot2)
+        picking2.move_line_ids.manual_lot_id = self.lot1
```

Comment:
You can set the lot here but because there is no unreserved quantity for this lot Odoo will reset the move line's product quantity.
```
-> self.assertEqual(picking2.move_line_ids.lot_id, self.lot2)
(Pdb) picking2.move_line_ids.read(["lot_id", "product_qty"])
[{'id': 53, 'lot_id': (41, 'lot2'), 'product_qty': 1.0}]
(Pdb) n
> /home/bruno/buildout/ou120/parts/stock-logistics-workflow/stock_move_manual_lot/tests/test_stock_move_manual_lot.py(68)test_reassign_reservation()
-> picking2.move_line_ids.manual_lot_id = self.lot1
(Pdb) 
> /home/bruno/buildout/ou120/parts/stock-logistics-workflow/stock_move_manual_lot/tests/test_stock_move_manual_lot.py(71)test_reassign_reservation()
-> with self.assertRaises(UserError), self.env.clear_upon_failure():
(Pdb) picking2.move_line_ids.read(["lot_id", "product_qty"])
[{'id': 53, 'lot_id': (40, 'lot1'), 'product_qty': 0.0}]
```
So I maintain you need to unreserve explicitely first.

Reply:
hmm, in my first naive version I called [_update_available_quantity(-this.product_qty)](https://github.com/OCA/OCB/blob/12.0/addons/stock/models/stock_quant.py#L189), but because [stock.move.line#write](https://github.com/OCA/OCB/blob/12.0/addons/stock/models/stock_move_line.py#L268) does so too, this breaks when we've only 2 on hand. Will investigate next week

Reply:
yes, we indeed need to [free the reservation](https://github.com/OCA/stock-logistics-workflow/pull/833/files#diff-8e597cc2aff6a170901caba6b907f585528d98ac5913bc38aa002775aa97a2dcR32) - the codepath in write [doing so](https://github.com/OCA/OCB/blob/12.0/addons/stock/models/stock_move_line.py#L308) only does it after it calculates the available quantity